File: week3/problem.java
class FizzBuzz {
    private int n;
    private int i=1;

    public FizzBuzz(int n) {
        this.n = n;
    }

    private void gate(Runnable printer, Supplier<Boolean> condition) throws InterruptedException {
        synchronized (this) {
            while(!condition.get() && i<=n) {
                wait();
            }
            if(i > n) return;
        }
        printer.run();
        synchronized (this) {
            i++;
            notifyAll();
        }
        gate(printer, condition);
    }

    // printFizz.run() outputs "fizz".
    public void fizz(Runnable printFizz) throws InterruptedException {
       gate(printFizz, () -> i%3 == 0 && i%5 != 0); 
    }

    // printBuzz.run() outputs "buzz".
    public void buzz(Runnable printBuzz) throws InterruptedException {
       gate(printBuzz, () -> i%3 != 0 && i%5 == 0); 
    }

    // printFizzBuzz.run() outputs "fizzbuzz".
    public void fizzbuzz(Runnable printFizzBuzz) throws InterruptedException {
       gate(printFizzBuzz, () -> i%3 == 0 && i%5 == 0); 
    }

    // printNumber.accept(x) outputs "x", where x is an integer.
    public void number(IntConsumer printNumber) throws InterruptedException {
       gate(() -> {printNumber.accept(i);}, () -> i%3 != 0 && i%5 != 0); 
    }

File: week3/problem.cpp
class FizzBuzz {
private:
    int n;
    int i = 1;
    mutex m; 
    condition_variable cv; 

public:
    FizzBuzz(int n) {
        this->n = n;
    }

    void gate(function<void()> func, function<bool()> condition) {
        {
            unique_lock lk(m);
            cv.wait(lk, [this,condition](){return condition() || i>n;});
            if(i > n) return;
        }
        func();
        {
            unique_lock lk(m);
            i++;
        }
        cv.notify_all();
        gate(func, condition);
    }

    // printFizz() outputs "fizz".
    void fizz(function<void()> printFizz) {
        gate(printFizz, [this](){return i%3==0 && i%5 != 0;});
    }

    // printBuzz() outputs "buzz".
    void buzz(function<void()> printBuzz) {
        gate(printBuzz, [this](){return i%3!=0 && i%5 == 0;});
    }

    // printFizzBuzz() outputs "fizzbuzz".
	void fizzbuzz(function<void()> printFizzBuzz) {
        gate(printFizzBuzz, [this](){return i%3==0 && i%5 == 0;});
    }

    // printNumber(x) outputs "x", where x is an integer.
    void number(function<void(int)> printNumber) {
        gate([this,printNumber](){printNumber(i);}, [this](){return i%3!=0 && i%5 != 0;});
    }
}
